File: sdk/typescript/src/crypto.ts
import { keccak256, encodePacked } from "viem";

/**
 * 生成 32 字节随机 salt
 */
export function generateSalt(): `0x${string}` {
  const bytes = new Uint8Array(32);
  crypto.getRandomValues(bytes);
  return `0x${Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("")}` as `0x${string}`;
}

/**
 * 生成 commit hash，与合约验证逻辑一致：
 * keccak256(abi.encodePacked(prediction, confidence, salt))
 */
export function makeCommitHash(
  prediction: number,
  confidence: number,
  salt: `0x${string}`
): `0x${string}` {
  return keccak256(
    encodePacked(
      ["uint8", "uint8", "bytes32"],
      [prediction, confidence, salt]
    )
  );
}

/**
 * 验证 commit hash
 */
export function verifyCommitHash(
  prediction: number,
  confidence: number,
  salt: `0x${string}`,
  expectedHash: `0x${string}`
): boolean {
  return makeCommitHash(prediction, confidence, salt) === expectedHash;
}

File: sdk/python/hive_protocol/crypto.py
"""commit-reveal 加密工具"""

import os
import struct
from eth_abi import encode
from web3 import Web3


def generate_salt() -> bytes:
    """生成 32 字节随机 salt"""
    return os.urandom(32)


def make_commit_hash(prediction: int, confidence: int, salt: bytes) -> bytes:
    """
    生成 commit hash，与合约中的验证逻辑一致：
    keccak256(abi.encodePacked(prediction, confidence, salt))

    prediction: 0=UP, 1=DOWN (uint8)
    confidence: 1-100 (uint8)
    salt: 32 bytes
    """
    packed = struct.pack("BB", prediction, confidence) + salt
    return Web3.keccak(packed)


def verify_commit_hash(prediction: int, confidence: int, salt: bytes, expected_hash: bytes) -> bool:
    """验证 commit hash"""
    actual = make_commit_hash(prediction, confidence, salt)
    return actual == expected_hash
